objects/oEnemy/Step_0.gml

```diff
@@ -1,74 +1,124 @@
 
-// Collision detection with the player and other enemies
-if (place_meeting(x + xSpeed, y, oEnemy) || place_meeting(x + xSpeed, y, oPlayer)) {
+			// Set up to account for +ve and -ve xSpeed
+			xIndex = floor(abs(xSpeed));
+			signXSpeed = sign(xSpeed);
+			for (var i = xIndex; i >= 0; i--) {
+				if (!(place_meeting(x + (i * signXSpeed), y, oEnemy) || place_meeting(x + (i * signXSpeed), y, oPlayer))) {
+					xSpeed = i * signXSpeed;
+					break;
+				}
+				xSpeed = 0;
+			}
+		}
+		if (place_meeting(x, y + ySpeed, oEnemy) || place_meeting(x, y + ySpeed, oPlayer)) {
 
-	// Set up to account for +ve and -ve xSpeed
-	xIndex = floor(abs(xSpeed));
-	signXSpeed = sign(xSpeed);
-	for (var i = xIndex; i >= 0; i--) {
-		if (!(place_meeting(x + (i * signXSpeed), y, oEnemy) || place_meeting(x + (i * signXSpeed), y, oPlayer))) {
-			xSpeed = i * signXSpeed;
-			break;
+			// Set up to account for +ve and -ve ySpeed
+			yIndex = floor(abs(ySpeed));
+			signYSpeed = sign(ySpeed);
+			for (var i = yIndex; i >= 0; i--) {
+				if (!(place_meeting(x, y + (i * signYSpeed), oEnemy) || place_meeting(x, y + (i * signYSpeed), oPlayer))) {
+					ySpeed = i * signYSpeed;
+					break;
+				}
+				xSpeed = 0;
+			}
 		}
-		xSpeed = 0;
-	}
-}
-if (place_meeting(x, y + ySpeed, oEnemy) || place_meeting(x, y + ySpeed, oPlayer)) {
 
-	// Set up to account for +ve and -ve ySpeed
-	yIndex = floor(abs(ySpeed));
-	signYSpeed = sign(ySpeed);
-	for (var i = yIndex; i >= 0; i--) {
-		if (!(place_meeting(x, y + (i * signYSpeed), oEnemy) || place_meeting(x, y + (i * signYSpeed), oPlayer))) {
-			ySpeed = i * signYSpeed;
-			break;
+		// Collision detection with the Enemy and walls
+		if (place_meeting(x + xSpeed, y, oDirtWall)) {
+			xSpeed = 0;
+		}
+		if (place_meeting(x, y + ySpeed, oDirtWall)) {
+			ySpeed = 0;
 		}
-		ySpeed = 0;
+		
+		// movement towards oPlayer
+		x += xSpeed;
+		y += ySpeed;
+		
+		
 	}
-}
-
-// Collision detection with the Enemy and walls
-if (place_meeting(x + xSpeed, y, oDirtWall)) {
-	xSpeed = 0;
-}
-if (place_meeting(x, y + ySpeed, oDirtWall)) {
-	ySpeed = 0;
-}
-
-// Snap to initial position
-if(distance_to_point(initial_x, initial_y) < walkSpeed/2 and chasing_player==false){
-	if(!place_meeting(initial_x,initial_y,oPlayer) && !place_meeting(initial_x,initial_y,oEnemy)){
-		x = initial_x;
-		y = initial_y;
+	else {
+		// Return to Path
+		speed = 0;
+		pathMode = 3; // Do nothing unless find a suitable return path
+		var returnx;
+		var returny;
+		var pathCheckPositions = [returnPathPosition, 0, 0.2, 0.4, 0.6, 0.8, 1];
+		var shortestDistance = room_height + room_width;
+		for(i=0; i<array_length_1d(pathCheckPositions); i++){
+			var checkPosition = pathCheckPositions[i];
+			returnx = path_get_x(patrol, checkPosition);
+			returny = path_get_y(patrol, checkPosition);
+			var currentDistance = point_distance(x, y, returnx, returny);
+			if(!collision_line( x, y, returnx, returny, oDirtWall, true, false ) && (currentDistance < shortestDistance)){
```

Comment:
I'm not sure why but the `collision_line()` function doesn't always seem to be working properly. Sometimes the enemy will path back through a wall, I can show you guys what I mean next time we're on Skype